File: gitjuggling/src/remote.rs
use anyhow::{Context, Result};
use serde::Deserialize;
use std::process::Command;

// ---------------------------------------------------------------------------
// RemoteRepo — common representation of a remote repository
// ---------------------------------------------------------------------------

#[derive(Debug, Clone)]
pub struct RemoteRepo {
    pub name: String,
    pub owner: String,
    pub is_fork: bool,
    pub is_archived: bool,
    #[allow(dead_code)]
    pub is_mirror: bool,
    #[allow(dead_code)]
    pub clone_url: String,
    #[allow(dead_code)]
    pub source: RemoteSource,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum RemoteSource {
    GitHub,
    Forgejo,
}

impl RemoteRepo {
    /// Unique key for deduplication: (owner, name) in lowercase.
    pub fn dedup_key(&self) -> (String, String) {
        (self.owner.to_lowercase(), self.name.to_lowercase())
    }
}

// ---------------------------------------------------------------------------
// GitHub client
// ---------------------------------------------------------------------------

#[derive(Debug, Deserialize)]
struct GitHubRepo {
    name: String,
    #[serde(default)]
    fork: bool,
    #[serde(default)]
    archived: bool,
    clone_url: String,
    owner: GitHubOwner,
}

#[derive(Debug, Deserialize)]
struct GitHubOwner {
    login: String,
}

/// Fetch a GitHub token using `gh auth token`.
fn github_token() -> Result<String> {
    let output = Command::new("gh")
        .args(["auth", "token"])
        .output()
        .context("failed to run `gh auth token`")?;

    if !output.status.success() {
        let stderr = String::from_utf8_lossy(&output.stderr);
        anyhow::bail!("`gh auth token` failed: {}", stderr.trim());
    }

    Ok(String::from_utf8(output.stdout)?.trim().to_string())
}

/// Fetch all repos for the given GitHub owners (users or orgs).
pub fn fetch_github_repos(owners: &[String]) -> Result<Vec<RemoteRepo>> {
    let token = github_token()?;
    let mut all_repos = Vec::new();

    for owner in owners {
        let repos = fetch_github_owner_repos(&token, owner)?;
        all_repos.extend(repos);
    }

    Ok(all_repos)
}

fn fetch_github_owner_repos(token: &str, owner: &str) -> Result<Vec<RemoteRepo>> {
    let client = reqwest::blocking::Client::new();
    let mut repos = Vec::new();
    let mut page = 1u32;

    loop {
        // Try org endpoint first, fall back to user endpoint
        let url = format!(
            "https://api.github.com/orgs/{}/repos?page={}&per_page=100&type=all",
            owner, page
        );

        let resp = client
            .get(&url)
            .header("Authorization", format!("Bearer {}", token))
            .header("User-Agent", "gitjuggling")
            .send()
            .context(format!(
                "failed to fetch GitHub repos for owner '{}'",
                owner
            ))?;

        // If 404, try user endpoint
        if resp.status() == reqwest::StatusCode::NOT_FOUND {
            let user_url = format!(
                "https://api.github.com/users/{}/repos?page={}&per_page=100&type=all",
                owner, page
            );

            let user_resp = client
                .get(&user_url)
                .header("Authorization", format!("Bearer {}", token))
                .header("User-Agent", "gitjuggling")
                .send()
                .context(format!("failed to fetch GitHub repos for user '{}'", owner))?;

            let page_repos: Vec<GitHubRepo> = user_resp.json()?;
            if page_repos.is_empty() {
                break;
            }

            for r in page_repos {
                repos.push(RemoteRepo {
                    name: r.name,
                    owner: r.owner.login,
                    is_fork: r.fork,
                    is_archived: r.archived,
                    is_mirror: false,
                    clone_url: r.clone_url,
                    source: RemoteSource::GitHub,
                });
            }

            page += 1;
            continue;
        }

        let page_repos: Vec<GitHubRepo> = resp.json()?;
        if page_repos.is_empty() {
            break;
        }

        for r in page_repos {
            repos.push(RemoteRepo {
                name: r.name,
                owner: r.owner.login,
                is_fork: r.fork,
                is_archived: r.archived,
                is_mirror: false,
                clone_url: r.clone_url,
                source: RemoteSource::GitHub,
            });
        }

        page += 1;
    }

    Ok(repos)
}

// ---------------------------------------------------------------------------
// Forgejo / Gitea client
// ---------------------------------------------------------------------------

#[derive(Debug, Deserialize)]
struct ForgejoRepo {
    name: String,
    #[serde(default)]
    fork: bool,
    #[serde(default)]
    archived: bool,
    #[serde(default)]
    mirror: bool,
    clone_url: String,
    owner: ForgejoOwner,
}

#[derive(Debug, Deserialize)]
struct ForgejoOwner {
    login: String,
}

/// Fetch a Forgejo token by executing the configured command.
fn forgejo_token(cmd: &str) -> Result<String> {
    let output = Command::new("sh")
        .arg("-c")
        .arg(cmd)
        .output()
        .context(format!("failed to execute forgejo token command: {}", cmd))?;

    if !output.status.success() {
        let stderr = String::from_utf8_lossy(&output.stderr);
        anyhow::bail!("forgejo token command failed ({}): {}", cmd, stderr.trim());
    }

    Ok(String::from_utf8(output.stdout)?.trim().to_string())
}

/// Fetch all repos for the given Forgejo user, filtering out mirrors.
pub fn fetch_forgejo_repos(base_url: &str, user: &str, token_cmd: &str) -> Result<Vec<RemoteRepo>> {
    let token = forgejo_token(token_cmd)?;
    let client = reqwest::blocking::Client::new();
    let mut repos = Vec::new();
    let mut page = 1u32;

    loop {
        let url = format!(
            "{}/api/v1/users/{}/repos?page={}&limit=50",
            base_url.trim_end_matches('/'),
            user,
            page
        );

        let resp = client
            .get(&url)
            .header("Authorization", format!("token {}", token))
            .send()
            .context(format!(
                "failed to fetch Forgejo repos for user '{}' at {}",
                user, base_url
            ))?;

        if !resp.status().is_success() {
            anyhow::bail!("Forgejo API returned {} for user '{}'", resp.status(), user);
        }

        let page_repos: Vec<ForgejoRepo> = resp.json()?;
        if page_repos.is_empty() {
            break;
        }

        for r in page_repos {
            // Skip mirrors — they are duplicates of GitHub repos
            if r.mirror {
                continue;
            }

            repos.push(RemoteRepo {
                name: r.name,
                owner: r.owner.login,
                is_fork: r.fork,
                is_archived: r.archived,
                is_mirror: r.mirror,
                clone_url: r.clone_url,
                source: RemoteSource::Forgejo,
            });
        }

        page += 1;
    }

    Ok(repos)
}

// ---------------------------------------------------------------------------
// Deduplication
// ---------------------------------------------------------------------------

/// Remove Forgejo repos that duplicate GitHub repos (same owner/name).
#[allow(clippy::ptr_arg)]
pub fn dedup_repos(github: &mut Vec<RemoteRepo>, forgejo: &mut Vec<RemoteRepo>) {
    let github_keys: std::collections::HashSet<(String, String)> =
        github.iter().map(|r| r.dedup_key()).collect();

    forgejo.retain(|r| !github_keys.contains(&r.dedup_key()));
}

File: gitjuggling/internal/remote/remote.go
package remote

import (
	"context"
	"encoding/json"
	"fmt"
	"net/http"
	"os/exec"
	"strings"
)

// RemoteSource indicates where a remote repo originates from.
type RemoteSource int

const (
	SourceGitHub RemoteSource = iota
	SourceForgejo
)

// RemoteRepo represents a repository from a remote source (GitHub or Forgejo).
type RemoteRepo struct {
	Name       string
	Owner      string
	IsFork     bool
	IsArchived bool
	IsMirror   bool
	CloneURL   string
	Source     RemoteSource
}

// DedupKey returns a lowercase (owner, name) tuple for deduplication.
func (r *RemoteRepo) DedupKey() [2]string {
	return [2]string{strings.ToLower(r.Owner), strings.ToLower(r.Name)}
}

// ---------------------------------------------------------------------------
// GitHub client
// ---------------------------------------------------------------------------

// githubRepo is the JSON response struct for GitHub API repos.
type githubRepo struct {
	Name     string `json:"name"`
	Fork     bool   `json:"fork"`
	Archived bool   `json:"archived"`
	CloneURL string `json:"clone_url"`
	Owner    struct {
		Login string `json:"login"`
	} `json:"owner"`
}

// GitHubToken retrieves a GitHub token using `gh auth token`.
func GitHubToken() (string, error) {
	cmd := exec.Command("gh", "auth", "token")
	out, err := cmd.Output()
	if err != nil {
		return "", fmt.Errorf("`gh auth token` failed: %w", err)
	}
	return strings.TrimSpace(string(out)), nil
}

// FetchGitHubRepos fetches all repos for the given GitHub owners (users or orgs).
func FetchGitHubRepos(owners []string) ([]*RemoteRepo, error) {
	token, err := GitHubToken()
	if err != nil {
		return nil, err
	}

	var allRepos []*RemoteRepo
	for _, owner := range owners {
		repos, err := fetchGitHubOwnerRepos(token, owner)
		if err != nil {
			return nil, fmt.Errorf("fetching GitHub repos for %q: %w", owner, err)
		}
		allRepos = append(allRepos, repos...)
	}
	return allRepos, nil
}

// fetchGitHubOwnerRepos paginates through the GitHub API for a single owner.
// It tries the org endpoint first, falling back to the user endpoint on 404.
func fetchGitHubOwnerRepos(token, owner string) ([]*RemoteRepo, error) {
	client := &http.Client{}
	var repos []*RemoteRepo
	page := 1

	for {
		// Try org endpoint first
		url := fmt.Sprintf("https://api.github.com/orgs/%s/repos?page=%d&per_page=100&type=all", owner, page)

		req, err := http.NewRequestWithContext(context.Background(), http.MethodGet, url, nil)
		if err != nil {
			return nil, err
		}
		req.Header.Set("Authorization", "Bearer "+token)
		req.Header.Set("User-Agent", "gitjuggling")

		resp, err := client.Do(req)
		if err != nil {
			return nil, fmt.Errorf("request to GitHub API failed: %w", err)
		}

		// If 404, try user endpoint
		if resp.StatusCode == http.StatusNotFound {
			resp.Body.Close()
			url = fmt.Sprintf("https://api.github.com/users/%s/repos?page=%d&per_page=100&type=all", owner, page)

			req, err = http.NewRequestWithContext(context.Background(), http.MethodGet, url, nil)
			if err != nil {
				return nil, err
			}
			req.Header.Set("Authorization", "Bearer "+token)
			req.Header.Set("User-Agent", "gitjuggling")

			resp, err = client.Do(req)
			if err != nil {
				return nil, fmt.Errorf("request to GitHub API failed: %w", err)
			}
		}
		defer resp.Body.Close()

		var pageRepos []githubRepo
		if err := json.NewDecoder(resp.Body).Decode(&pageRepos); err != nil {
			return nil, fmt.Errorf("decoding GitHub response: %w", err)
		}

		if len(pageRepos) == 0 {
			break
		}

		for _, r := range pageRepos {
			repos = append(repos, &RemoteRepo{
				Name:       r.Name,
				Owner:      r.Owner.Login,
				IsFork:     r.Fork,
				IsArchived: r.Archived,
				IsMirror:   false,
				CloneURL:   r.CloneURL,
				Source:     SourceGitHub,
			})
		}

		page++
	}

	return repos, nil
}

// ---------------------------------------------------------------------------
// Forgejo / Gitea client
// ---------------------------------------------------------------------------

// forgejoRepo is the JSON response struct for Forgejo API repos.
type forgejoRepo struct {
	Name     string `json:"name"`
	Fork     bool   `json:"fork"`
	Archived bool   `json:"archived"`
	Mirror   bool   `json:"mirror"`
	CloneURL string `json:"clone_url"`
	Owner    struct {
		Login string `json:"login"`
	} `json:"owner"`
}

// httpsToSSH converts an HTTPS git URL to SSH format.
//
// Examples:
//
//	https://git.example.com/user/project.git → git@git.example.com:user/project.git
//	https://git.example.com/user/project      → git@git.example.com:user/project.git
func httpsToSSH(url string) string {
	url = strings.TrimRight(url, "/")

	if after, ok := strings.CutPrefix(url, "https://"); ok {
		after = strings.TrimSuffix(after, ".git")
		host, path, ok := strings.Cut(after, "/")
		if !ok {
			return url // can't parse, return as-is
		}
		return "git@" + host + ":" + path + ".git"
	}

	return url // not HTTPS, return as-is
}

// ForgejoToken retrieves a Forgejo token by reading the given 1Password secret reference.
// The secretRef should be an "op://..." URI; this function hardcodes "op read" to execute it.
func ForgejoToken(secretRef string) (string, error) {
	out, err := exec.Command("op", "read", secretRef).Output()
	if err != nil {
		return "", fmt.Errorf("op read failed for %q: %w", secretRef, err)
	}
	return strings.TrimSpace(string(out)), nil
}

// FetchForgejoRepos fetches all repos for the given Forgejo user, filtering out mirrors.
func FetchForgejoRepos(baseURL, user, tokenRef string) ([]*RemoteRepo, error) {
	token, err := ForgejoToken(tokenRef)
	if err != nil {
		return nil, err
	}

	client := &http.Client{}
	var repos []*RemoteRepo
	page := 1

	baseURL = strings.TrimRight(baseURL, "/")

	for {
		url := fmt.Sprintf("%s/api/v1/users/%s/repos?page=%d&limit=50", baseURL, user, page)

		req, err := http.NewRequestWithContext(context.Background(), http.MethodGet, url, nil)
		if err != nil {
			return nil, err
		}
		req.Header.Set("Authorization", "token "+token)

		resp, err := client.Do(req)
		if err != nil {
			return nil, fmt.Errorf("request to Forgejo API failed for user %q at %s: %w", user, baseURL, err)
		}
		defer resp.Body.Close()

		if resp.StatusCode != http.StatusOK {
			return nil, fmt.Errorf("Forgejo API returned %d for user %q", resp.StatusCode, user)
		}

		var pageRepos []forgejoRepo
		if err := json.NewDecoder(resp.Body).Decode(&pageRepos); err != nil {
			return nil, fmt.Errorf("decoding Forgejo response: %w", err)
		}

		if len(pageRepos) == 0 {
			break
		}

		for _, r := range pageRepos {
			// Skip mirrors — they are duplicates of GitHub repos
			if r.Mirror {
				continue
			}

			repos = append(repos, &RemoteRepo{
				Name:       r.Name,
				Owner:      r.Owner.Login,
				IsFork:     r.Fork,
				IsArchived: r.Archived,
				IsMirror:   r.Mirror,
				CloneURL:   httpsToSSH(r.CloneURL),
				Source:     SourceForgejo,
			})
		}

		page++
	}

	return repos, nil
}

// ---------------------------------------------------------------------------
// Deduplication
// ---------------------------------------------------------------------------

// DedupRepos removes Forgejo repos that duplicate GitHub repos (same owner/name).
// It returns the modified github and forgejo slices.
func DedupRepos(github, forgejo []*RemoteRepo) ([]*RemoteRepo, []*RemoteRepo) {
	githubKeys := make(map[[2]string]struct{}, len(github))
	for _, r := range github {
		githubKeys[r.DedupKey()] = struct{}{}
	}

	var filtered []*RemoteRepo
	for _, r := range forgejo {
		if _, dup := githubKeys[r.DedupKey()]; !dup {
			filtered = append(filtered, r)
		}
	}

	return github, filtered
}
